File: LeetCode/Python/Medium/276-Paint-Fence/solution.py
'''
Solution 1: pure recursion. Note: this solution will get time limit exceeded when submitted. DO NOT submit this one!!!
'''
class Solution:
    def numWays(self, n: int, k: int) -> int:
        result=0

        previous_color=-1

        previous_two_fences_the_same_color=0

        result=self.helper(n, k, previous_color, previous_two_fences_the_same_color)

        return result

    def helper(self, n: int, k: int, previous_color: int, previous_two_fences_the_same_color: int) -> int:
        if n==0:
            return 1

        subproblem_solution=0

        for next_color in range(0, k):
            if previous_color==next_color and previous_two_fences_the_same_color:
                continue

            subproblem_solution+=self.helper(n - 1, k, next_color, int(previous_color==next_color))

        return subproblem_solution

'''
Solution 2: recursion + memoization. The online judge accepts this solution.
'''
class Solution:
    def numWays(self, n: int, k: int) -> int:
        result=0

        previous_color=-1

        previous_two_fences_the_same_color=0

        memo=[[-1 for _ in range(0, 2)] for _ in range(0, n + 1)]

        result=self.helper(memo, n, k, previous_color, previous_two_fences_the_same_color)

        return result

    def helper(self, memo: List[List[int]], n: int, k: int, previous_color: int, previous_two_fences_the_same_color: int) -> int:
        if n==0:
            return 1

        if memo[n][previous_two_fences_the_same_color]!=-1:
            return memo[n][previous_two_fences_the_same_color]

        subproblem_solution=0

        for next_color in range(0, k):
            if previous_color==next_color and previous_two_fences_the_same_color:
                continue

            subproblem_solution+=self.helper(memo, n - 1, k, next_color, int(previous_color==next_color))

        memo[n][previous_two_fences_the_same_color]=subproblem_solution

        return subproblem_solution

File: LeetCode/Rust/Medium/276-Paint-Fence/solution.rs
impl Solution
{
    pub fn num_ways(n: i32, k: i32) -> i32
    {
        let mut result: i32=0;

        let mut memo: Vec<Vec<i32>>=vec![vec![-1; 2]; (n + 1) as usize];

        let mut previous_color: i32=-1;

        let previous_two_posts_the_same_color: i32=0;

        result=Self::helper(&mut memo, n, k, previous_color, previous_two_posts_the_same_color);

        result
    }

    fn helper(memo: &mut Vec<Vec<i32>>, n: i32, k: i32, previous_color: i32, previous_two_posts_the_same_color: i32) -> i32
    {
        if n==0
        {
            return 1;
        }

        if memo[n as usize][previous_two_posts_the_same_color as usize]!=-1
        {
            return memo[n as usize][previous_two_posts_the_same_color as usize];
        }

        let mut subproblem_solution: i32=0;

        for next_color in 0..k
        {
            if previous_color==next_color && previous_two_posts_the_same_color==1
            {
                continue;
            }

            subproblem_solution+=Self::helper(memo, n - 1, k, next_color, (previous_color==next_color) as i32);
        }

        memo[n as usize][previous_two_posts_the_same_color as usize]=subproblem_solution;

        return subproblem_solution;
    }
}